src/NHibernate/Transaction/AdoNetWithDistributedTransactionFactory.cs

```diff
@@ -178,6 +163,22 @@ void IEnlistmentNotification.Prepare(PreparingEnlistment preparingEnlistment)
 				}
 			}
 
+			static void BeforeTransactionCompletion(ISessionImplementor session)
+			{
+				if (session.ConnectionManager.IsConnected)
+				{
+					using (session.ConnectionManager.FlushingFromDtcTransaction)
```

Comment:
Calling this in a loop could be later a trouble, if we actually merge the connection switch it causes in [#627](https://github.com/nhibernate/nhibernate-core/pull/627) for database supporting it, in order to prevent race condition between the connection own 2PC and NHibernate 2PC. I would rather have it surrounding the whole loop.